File: 2022.01.31/javascript/10451.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');

const search = (idx,visited,arr) =>{
  if (visited[idx]) return 1;
  visited[idx] = true;
  return search(arr[idx-1],visited,arr);
}
for ( let i = 1 ; i<input.length-1 ; i+=2){
  const n = Number(input[i]);
  const arr = input[i+1].split(' ').map(Number)
  let visited = new Array(n+1).fill(false);
  let result = 0;
  for (let j = 1 ; j<=n ; j++){
    if (visited[j]) continue;
    result += search(j,visited,arr);
  }
  console.log(result);
}

File: 2022.01.31/python/10451.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
t = int(input())
def search(num):
    if visited[num] : return 1
    visited[num] = True
    return search(nums[num])

for i in range(t):
    n = int(input())
    nums = [0] +  list(map(int,input().split()))
    visited = [False for _ in range(n+1)]
    result = 0
    for i in range(1,len(nums)):
        if visited[i]: continue
        result += search(i)
    print(result)